code/modules/admin/smites/puzzle.dm

```diff
@@ -0,0 +1,330 @@
+		return -1
+	return 0
+
+/obj/effect/sliding_puzzle/proc/elements_in_order()
+	return sortTim(elements,cmp=/proc/cmp_xy_desc)
+
+/obj/effect/sliding_puzzle/proc/get_base_icon()
+	var/icon/I = new('icons/obj/puzzle.dmi')
+	var/list/puzzles = icon_states(I)
+	var/puzzle_state = pick(puzzles)
+	var/icon/P = new('icons/obj/puzzle.dmi',puzzle_state)
+	return P
+
+/obj/effect/sliding_puzzle/proc/setup()
+	//First we slice the 96x96 icon into 32x32 pieces
+	var/list/puzzle_pieces = list() //id -> icon list
+
+	var/width = 3
+	var/height = 3
+	var/list/left_ids = list()
+	var/tile_count = width * height
+
+	//Generate per tile icons
+	var/icon/base_icon = get_base_icon()
+
+	for(var/id in 1 to tile_count)
+		var/y = width - round((id - 1) / width)
+		var/x = ((id - 1) % width) + 1
+
+		var/x_start = 1 + (x - 1) * world.icon_size
+		var/x_end = x_start + world.icon_size - 1
+		var/y_start = 1 + ((y - 1) * world.icon_size)
+		var/y_end = y_start + world.icon_size - 1
+
+		var/icon/T = new(base_icon)
+		T.Crop(x_start,y_start,x_end,y_end)
+		puzzle_pieces["[id]"] = T
+		left_ids += id
+
+	//Setup random empty tile
+	empty_tile_id = pick_n_take(left_ids)
+	var/turf/empty_tile_turf = get_turf_for_id(empty_tile_id)
+	empty_tile_turf.PlaceOnTop(floor_type,null)
+	var/mutable_appearance/MA = new(puzzle_pieces["[empty_tile_id]"])
+	MA.layer = empty_tile_turf.layer + 0.1
+	empty_tile_turf.add_overlay(MA)
+
+	elements = list()
+	var/list/empty_spots = left_ids.Copy()
+	for(var/spot_id in empty_spots)
+		var/turf/T = get_turf_for_id(spot_id)
+		T = T.PlaceOnTop(floor_type,null)
+		var/obj/structure/puzzle_element/E = new element_type(T)
+		elements += E
+		var/chosen_id = pick_n_take(left_ids)
+		E.puzzle_icon = puzzle_pieces["[chosen_id]"]
+		E.source = src
+		E.id = chosen_id
+		E.set_puzzle_icon()
+
+	if(!is_solvable())
+		make_solvable()
+
+/obj/structure/puzzle_element
+	name = "mysterious pillar"
+	desc = "puzzling..."
+	icon = 'icons/obj/artefacts.dmi'
+	icon_state = "puzzle_pillar"
+	anchored = FALSE
+	density = TRUE
+	var/id = 0
+	var/obj/effect/sliding_puzzle/source
+	var/icon/puzzle_icon
+
+/obj/structure/puzzle_element/Move(nloc, dir)
+	if(!isturf(nloc) ||  moving_diagonally || get_dist(get_step(src,dir),get_turf(source)) > 1)
+		return 0
+	else
+		return ..()
+
+/obj/structure/puzzle_element/proc/set_puzzle_icon()
+	cut_overlays()
+	if(puzzle_icon)
+		//Need to scale it down a bit to fit the static border
+		var/icon/C = new(puzzle_icon)
+		C.Scale(19,19)
+		var/mutable_appearance/puzzle_small = new(C)
+		puzzle_small.layer = layer + 0.1
+		puzzle_small.pixel_x = 7
+		puzzle_small.pixel_y = 7
+		add_overlay(puzzle_small)
+
+/obj/structure/puzzle_element/Destroy()
+	if(source)
+		source.elements -= src
+		source.validate()
+	return ..()
+
+//Set the full image on the turf and delete yourself
+/obj/structure/puzzle_element/proc/collapse()
+	var/turf/T = get_turf(src)
+	var/mutable_appearance/MA = new(puzzle_icon)
+	MA.layer = T.layer + 0.1
```

Comment:
bad var name.
var name should be immediatly understandable, so not MA, but var/mutable_appearance/mutable_appearance =